refactor(compiler): generate for loop block instructions

crisbeto opened the pull request:
Adds the initial implementation to generate the instructions for the `for` loop block.

**Note:** the expressions we support in the `track` paramateter are currently limited to tracking by identity or index, or a specific property of the item. Supporting more advanced expression will require additional work that I'll do in a follow-up PR.

crisbeto commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -913,61 +926,67 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
-      sourceSpan: ParseSourceSpan, variables: t.Variable[] = [], attrsExprs?: o.Expression[],
-      references?: t.Reference[], i18n?: i18n.I18nMeta): number {
-    const templateIndex = this.allocateDataSlot();
+  private prepareEmbeddedTemplateFn(
```
Context here: I had to split up the `createEmbeddedTemplateFn` method into two, because we needed the ability to create a template function without generating an instruction for it. Instead we take the parameters and we pass them into the `repeaterCreate` instruction which calls the `template` instruction internally.

The gotcha here is that we need to read const and declaration count after all the child templates have been processed. That's why I added a the helper `TemplateData` class.

crisbeto commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -15,6 +15,15 @@ import {getLView, nextBindingIndex} from '../state';
 import {addLViewToLContainer, createAndRenderEmbeddedLView, getLViewFromLContainer, removeLViewFromLContainer} from '../view_manipulation';
 
+// TODO: runtime repeater implementation
+export function ɵɵrepeaterCreate(
```
I copied the signatures for these instructions from #51422.

AndrewKushnir commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_control_flow/for_aliased_template_variables_template.js:
```diff
@@ -0,0 +1,24 @@
+  if (rf & 2) {
+    const idx_r2 = ctx.$index;
+    const co_r3 = ctx.$count;
+    $r3$.ɵɵtextInterpolate6(" Index: ", idx_r2, " First: ", idx_r2 === 0, " Last: ", idx_r2 === co_r3 - 1, " Even: ", idx_r2 % 2 === 0, " Odd: ", idx_r2 % 2 !== 0, " Count: ", co_r3, " ");
```
Quick question: what happens if there is a code like this in a template: `{{ $first + 1 }}` (there might be better cases where built-in local vars are used in expressions)? Would that be turned into `idx_r2 === 0 + 1` (which would affect the outcome)? I'm wondering if in this case we can extract `idx_r2 === 0` (and others) to a separate var or wrap them in parens if an expressions is more complex than just `{{ $first }}`?

crisbeto replied:
Good point 👍 I would expect the output AST to handle it. Also in template type checking we would treat it as a boolean which will result in an error. I'll add a test for it.

crisbeto replied:
@AndrewKushnir I've added the test. It looks like the output AST was accounting for it already.

AndrewKushnir commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_control_flow/for_with_empty.ts:
```diff
@@ -0,0 +1,17 @@
+      {#for item of items; track item}
+        {{item.name}}
+        {:empty} No items!
+      {/for}
```
I'd propose to update one of the tests with an `:empty` placeholder to add an element after the `for` block, so that we can verify that we correctly allocate an extra slot in `TView.data` for the empty template.

crisbeto replied:
Yeah, that's a good call. I noticed that the index was wrong when I removed the manually-written instructions from the runtime tests. I've added a test that has an `ng-template` before and after the `for` block.

AndrewKushnir reviewed (approved):
LGTM, just left a minor comment (+ CI seems to fail with a legit error).

crisbeto committed:
refactor(compiler): generate for loop block instructions

Adds the initial implementation to generate the instructions for the `for` loop block.

**Note:** the expressions we support in the `track` paramateter are currently limited to tracking by identity or index, or a specific property of the item. Supporting more advanced expression will require additional work that I'll do in a follow-up PR.

crisbeto committed:
test(core): convert for block tests to template

Reworks the `for` block tests to use compiled templates instead of manually-written instructions.

crisbeto commented:
Pushed an extra commit that converts the runtime tests from manually-written instructions to compiled code.

jessicajaniuk commented:
This PR was merged into the repository by commit 4dabcad557a41d6f2e9542cb250990221fd36c22.
